segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StorageHealthTracker.java

```diff
@@ -0,0 +1,316 @@
+     */
+    private final AtomicBoolean isStorageDegraded = new AtomicBoolean(false);
+
+    /**
+     * Tracks whether storage should be throttled.
+     */
+    private final AtomicBoolean shouldThrottle = new AtomicBoolean(false);
+
+    /**
+     * Tracks whether storage is unavailable.
+     */
+    private final AtomicBoolean isStorageUnavailable = new AtomicBoolean(false);
+
+    /**
+     * Tracks whether storage is full.
+     */
+    private final AtomicBoolean isStorageFull = new AtomicBoolean(false);
+
+    /**
+     * Tracks percentage late requests in last iteration.
+     */
+    private final AtomicDouble percentageLate = new AtomicDouble(0);
+
+    /**
+     * Tracks storage used in bytes.
+     */
+    private final AtomicLong storageUsed = new AtomicLong(0);
+
+    /**
+     * Tracks number of late requests in current iteration.
+     */
+    private final AtomicLong lateRequestCount = new AtomicLong();
+
+    /**
+     * Tracks number of late requests in current iteration.
+     */
+    private final AtomicLong completedRequestCount = new AtomicLong();
+
+    /**
+     * Tracks number of late requests in current iteration.
+     */
+    private final AtomicLong pendingRequestCount = new AtomicLong();
+
+    /**
+     * Tracks number of unavailable requests in current iteration.
+     */
+    private final AtomicLong unavailableRequestCount = new AtomicLong();
+
+    /**
+     * Container Id.
+     */
+    private final long containerId;
+
+    /**
+     * {@link ChunkedSegmentStorageConfig} instance to use.
+     */
+    @NonNull
+    @Getter
+    private final ChunkedSegmentStorageConfig config;
+
+    /**
+     * Function that supplies current time.
+     */
+    @NonNull
+    private final Supplier<Long> currentTimeSupplier;
+
+    /**
+     * Function that supplies delay future.
+     */
+    @NonNull
+    private final Function<Duration, CompletableFuture<Void>> delaySupplier;
+
+    /**
+     * Indicates a beginning of the iteration or batch of operations.
+     *
+     * @param iterationId identifier of iteration.
+     */
+    public void beginIteration(long iterationId) {
+        // Clear
+        lateRequestCount.set(0);
+        completedRequestCount.set(0);
+        unavailableRequestCount.set(0);
+        lateRequestCount.set(0);
+    }
+
+    /**
+     * Indicates an end of the iteration or batch of operations.
+     *
+     * @param iterationId identifier of iteration.
+     */
+    public void endIteration(long iterationId) {
+        // Set the new percentage
+        if (completedRequestCount.get() == 0) {
+            percentageLate.set(0);
+        } else {
+            percentageLate.set((100.0 * lateRequestCount.get()) / completedRequestCount.get());
+        }
+
+        // Set throttle status
+        if (percentageLate.intValue() >= config.getMaxLateThrottlePercentage()) {
+            log.info("StorageHealthTracker[{}]: Storage is slow. {}% Requests are slow. Max {}% Min {}% allowed.",
```

Comment:
These log lines may be printed continuously if some Storage requests are late. This may flood our logs. Please, consider making all these to `debug` or just print the log when the state transition happens (e.g., from normal to late, or late to degraded or any of the previous to normal again).

Reply:
fixed